server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -100,73 +105,107 @@ class ObjectService(
     }
   } yield res
 
-  def checkMetadata(req: GetMetadataRequest): IO[MetadataResponse] =
-    for {
-      traceId <- IO(TraceId(randomUUID()))
-      storageLink <- findStorageLink(req.localObjectPath)
-      res <- storageLink.fold[IO[MetadataResponse]](IO.raiseError(StorageLinkNotFoundException(s"No storage link found for ${req.localObjectPath}"))) { pair =>
-        val isSafeMode = pair._1
-        val sl = pair._2
-        if(isSafeMode)
-          IO.pure(MetadataResponse.SafeMode(sl))
-        else {
-          for {
-            fullBlobName <- IO.fromEither(getFullBlobName(req.localObjectPath, sl.cloudStorageDirectory.blobPath).leftMap(s => BadRequestException(s)))
-            metadata <- retrieveGcsMetadata(req.localObjectPath, sl.cloudStorageDirectory.bucketName, fullBlobName, traceId)
-            result <- metadata match {
-              case None =>
-                IO(MetadataResponse.RemoteNotFound(sl))
-              case Some(GcsMetadata(_, lastLockedBy, expiresAt, crc32c, generation)) =>
-                val localAbsolutePath = config.workingDirectory.resolve(req.localObjectPath)
-                for {
-                  fileBody <- io.file.readAll[IO](localAbsolutePath, blockingEc, 4096).compile.to[Array]
-                  calculatedCrc32c = Crc32c.calculateCrc32c(fileBody)
-                  syncStatus <- if (calculatedCrc32c == crc32c) IO.pure(SyncStatus.Live) else {
-                    for {
-                      cachedGeneration <- metadataCache.get.map(_.get(req.localObjectPath))
-                      status <- cachedGeneration match {
-                        case Some(cachedGen) => if(cachedGen.generation == generation) IO.pure(SyncStatus.LocalChanged) else IO.pure(SyncStatus.RemoteChanged)
-                        case None => IO.pure(SyncStatus.OutOfSync)
-                      }
-                    } yield status
-                  }
-                  currentTime <- timer.clock.realTime(TimeUnit.MILLISECONDS)
-                  lastLock <- expiresAt.flatTraverse[IO, WorkbenchEmail] { ea =>
-                    if (currentTime > ea.toEpochMilli)
-                      IO.pure(none[WorkbenchEmail])
-                    else
-                      IO.pure(lastLockedBy)
-                  }
-                } yield MetadataResponse.EditMode(syncStatus, lastLock, generation, sl)
-            }
-          } yield result
-        }
+  def checkMetadata(req: GetMetadataRequest): Kleisli[IO, TraceId, MetadataResponse] = {
+    findStorageLink(
+      req.localObjectPath,
+      pair => Kleisli.liftF(IO.pure(MetadataResponse.SafeMode(pair.storageLink))),
```

Comment:
i think people who aren't cats-savvy would benefit from a brief comment describing what `Kleisli` is and why it's used here.

in general i do worry about how FP-heavy welder is and how it could be challenging for somebody to pick it up and debug welder issues. this is probably a larger discussion but i think the more comments the better. this is a small repo, so i think it would be a good model-repo for showing off some of these FP approaches and comments would go a long way.